Troubleshooting section for the ROS 2 simulation (#2353)
o3de/o3de.org: content/docs/user-guide/interactivity/robotics/troubleshooting.md

---
linkTitle: Troubleshooting
title: Troubleshooting
description: Troubleshooting common issues with Robot Operating System (ROS) and and the ROS 2 Gem in Open 3D Engine (O3DE).
weight: 500
---

In this section, you will learn helpful hints and what to check when your robotics simulation in Open 3D Engine (O3DE) is not working as expected.
You will also find solutions for some of the most commonly occurring problems.

> 

Reviewer: This is a good way to let customers know where to reach out for more help. 

```suggestion

If you don't find your problem covered here, try searching the issues and discussions in [`o3de-extras`](https://github.com/o3de/o3de-extras) repo or asking in the `#sig-simulation` channel of the [O3DE Discord](https://{{< links/o3de-discord >}}).

```

> ## A quick checklist

Reviewer: Suggestion to rename both level 2 headers (##) so that they go well together: 
- Is it a ROS 2 issue?
- Is it a Gem or O3DE issue?
```suggestion
## Is it a ROS 2 issue?
```

  - `echo $ROS_DISTRO` should show non-empty value, for example `humble`.
  - `echo $AMENT_PREFIX_PATH` should include your ROS 2 distribution installation path as well as any additional workspaces you have sourced (if any). 
  - If you are using ROS services in your project, make sure that the `RMW_IMPLEMENTATION` environment variable is the same on the both ends (check in each).

#### Node and topic visibility

If your simulation is running, you should see both ROS node(s) and topics.
  - Run `ros2 node list`. You should see at least `/o3de_ros2_node`.
  - Run `ros2 topic list` to list the topics. You should always see `/parameters` and `/rosout`.
    - You should see additional topics such as `/clock` and `/tf` if your simulation is running.

#### Message traffic

- Is there traffic on ROS 2 topics? When your simulation is running, messages should be published.
  - Check `ros2 topic hz` or `ros2 topic echo`. If you are seeing no traffic it could be caused by a firewall, disabled multicast or issues with your docker (if running from a docker).
  - Please refer to the [Troubleshooting guide](#troubleshooting-guide).
  

> #### Troubleshooting guide

Reviewer: Added "ROS 2" for clarity. Without it, this section title may be too ambiguous, seeing as this page is also a troubleshooting guide. 
```suggestion
#### ROS 2 troubleshooting guide
```

> See the [ROS 2 troubleshooting guide](https://docs.ros.org/en/rolling/How-To-Guides/Installation-Troubleshooting.html) for solutions.

Reviewer: Rewording suggestion. Changed wording to the possessive form, "ROS 2's", and updated the link to be the exact title of ROS 2's docs page. The motivation of this change is to make it clearer that we are linking to docs specifically in ROS 2's documentation set, not ours. 
```suggestion
For additional solution, refer to ROS 2's [Installation troubleshooting](https://docs.ros.org/en/rolling/How-To-Guides/Installation-Troubleshooting.html) page.
```

> ## It is a Gem or O3DE issue

Reviewer: Rephrased as a question. Also see comment on line 11 for the first ## header.
```suggestion
## Is it a Gem or O3DE issue?
```

> If your debugging confirmed that the issue is either with the ROS 2 Gem or O3DE, please help the community by raising an issue.

Reviewer: Added links to repos
```suggestion
If your debugging confirm that the issue is with either the ROS 2 Gem or O3DE, please help the community by raising an issue in [`o3de-extras`](https://github.com/o3de/o3de-extras/issues) or [`o3de`](https://github.com/o3de/o3de/issues) repos. First, please check the list of reported issues to avoid duplicates. 
```
